### DISCUSSION TITLE
Hardcoded carrier names

### DISCUSSION THREAD

**Discussion by drdanz** [robotology/yarp#1592] [General]
YARP is supposed to be agnostic about which carrier is currently used, but in some places the carrier name is checked. For example:

https://github.com/robotology/yarp/blob/290545d37e59d5d8ae3f1c55df151c656b308152/src/libYARP_OS/src/NameServer.cpp#L367-L377

In this example `local` and `shmem` are checked explicitly, but this will not work for any other carrier.
I think we should add 2 methods to the `Carrier` interface that allows to perform this check, something like `sameProcessOnly()`, and `sameSystemOnly()` both returning to false if not explicitly implemented in the carrier (Any suggestion on the naming is welcome).

I'm not really sure if this is ok though, since this part will run (I think) on the yarpserver, and it could be running on a machine where the carrier is not available. For example imagine a that we had a carrier that runs only on windows, and the yarp server is running on linux... How can we handle this?

There are other hardcoded carrier names (mostly `tcp`), we should also investigate those.

**Comment by Nicogene** [comment]
Unfortunately there is not the 😭 reaction. :grimacing: 

I totally agree to remove these hardcoded checks and define functions to the carrier level. 

In another carrier  that I worked to, I do this check inside the carrier itself during the initial handshake.
So if two ports are not on the same machine the connect fails, I don't know if it is possible to do the same for the `local` and `shmem` carrier.

**Comment by lornat75** [comment]
what about:

`interProcess()`/`intraProcess()` and `interHost()`/`intraHost()`

**Comment by drdanz** [comment]
Sounds good, but I always have doubts about `intra` vs `inter` :persevere:

**Comment by drdanz** [comment]
After a quick search, I'd say that `intraProcess()` and `intraHost()` are the right ones...

**Comment by drdanz** [comment]
I think we should also specify in the documentation the exact meaning of these, because they are already confusing to me...

`intraProcess()` should be "works intra process only" (i.e. only `local`) or "works intra process" (i.e. all carriers)
`intraHost()` should be "works intra host only" (i.e. `shmem`) or "works intra host" (i.e. all carriers, except `local`)

**Comment by lornat75** [comment]
Right. Is there the need to also specify that a protocol is `interHost` ?

**Comment by drdanz** [comment]
```c++
// Carrier.h

/**
 * Returns true if this carrier works only between 2 ports on the same process
 */
bool interProcessOnly() = 0; // { return false; } in AbstractCarrier, reimplemented in local carrier

/**
 * Returns true if this carrier works only between 2 ports on the same host
 */
bool interHostOnly() = 0; // { return false; } in AbstractCarrier, reimplemented in shmem carrier
```

Or perhaps something like this would be better

```c++
enum class Foo
{
    None = 0,
    IntraProcess,
    IntraHost,
    Any
};

Foo aSignificantMethodName()
{
    return Foo::Any;
}
```

or also (over-engineering?)

```c++
enum class Foo
{
    None = 0,
    WorksIntraProcess = 0x0001,
    WorksIntraHost = 0x0002,
    WorksInterHost = 0x0004,

    IntraProcessOnly = WorksIntraProcess,
    IntraHostOnly = WorksIntraProcess | WorksIntraHost,
    Any = WorksIntraProcess | WorksIntraHost | WorksInterHost
};

Foo aSignificantMethodName()
{
    return Foo::Any;
}
```

**Comment by barbalberto** [comment]
Personally, I'd prefer the first option. More clear and I get directly the info I want.
Btw, I think the name should be `intraXXX` instead of `interXXX`.

> Inter- is a prefix that means between two groups, and intra- is a prefix which means within or inside one group.

Or maybe `SameProcessOnly` `SameHostOnly` :smile:

**Comment by lornat75** [comment]
The second option has the advantage that it makes the options mutually exclusive, whereas in the first case you have to explicitly make sure that a carrier does not return true on both.

**Comment by barbalberto** [comment]
I was wandering exactly about that.
A intra-host carrier could also be intra-process, in theory, while it is not true the other way around.

**Comment by drdanz** [comment]
As discussed with @lornat75, we are proposing to use this naming:

```c++
enum class Foo
{
    None = 0,
    
    WorksIntraProcess = 0x0001,
    WorksIntraHost    = 0x0002,
    WorksInterHost    = 0x0004,

    IntraProcess = WorksIntraProcess
    IntraHostOnly = WorksIntraHost,
    IntraHost = WorksIntraProcess | WorksIntraHost,
    InterHostOnly = WorksInterHost,
    InterHost =  WorksIntraProcess | WorksIntraHost | WorksInterHost
};
